### ISSUE TITLE
Support patching NDPI files

### ISSUE THREAD

**Issue by erikogabrielsson** [cgohlke/tifffile#137]
Hi Christoph,

When trying to use the overwrite()-method on a tag in a ndpi-file this code raises an exception:
```python 
if tiff.version == 42 and tiff.offsetsize == 8:
    # TODO: support patching NDPI > 4 GB files
    raise NotImplementedError('cannot patch NDPI > 4 GB files')
``` 
Is this of concern even if the file is < 4 GB? Or if the size of the new value is not larger than the old? What is needed to implement patching ndpi files?

Regards

**Comment by cgohlke** [comment]
> Is this of concern even if the file is < 4 GB?

If the file size stays below 4 GB and LONG type values do not exceed 32-bit, it might work. Try to rename the file extension of a NDPI file < 4 GB to '.tif' and patch the tags, then rename the file back to its original. I should add an option to the TiffFile constructor so the renaming is not necessary...

> What is needed to implement patching ndpi files?

Ideally, a specification of the NDPI file format. The current understanding of the format is based on reverse engineering and bits of information found on the Internet. 
For files > 4 GB, the higher 4 bytes of 64-bit tag value offsets have to be written to a separate area after the IFD structure, equivalent to the code for reading:

https://github.com/cgohlke/tifffile/blob/375d97f62df6482142b51f1b38a49bdd24d18a60/tifffile/tifffile.py#L6604-L6611

I'll try to implement it.

Still, without a specification you are on thin ice. The value of tags in NDPI files is sometimes obscure, e.g.

https://github.com/cgohlke/tifffile/blob/375d97f62df6482142b51f1b38a49bdd24d18a60/tifffile/tifffile.py#L5415-L5419

**Comment by cgohlke** [comment]
Also, beware that many tag values like Software, Make, Model, and DateTime are shared between pages, which is not TIFF conformant. If overwriting one of those tags changes the offset of the value in the file, you'll need to manually adjust the tag value offsets in the other pages. There is not high level support for that in tifffile.

**Comment by erikogabrielsson** [comment]
Thanks,
I have had some success with replacing values with same length-values, i.e. this case:
 https://github.com/cgohlke/tifffile/blob/375d97f62df6482142b51f1b38a49bdd24d18a60/tifffile/tifffile.py#L9376-L9383

However, I think the format for packing the count is wrong? For offsetformat is set to '<Q' but the count should formatted as '<I'?

**Comment by cgohlke** [comment]
> However, I think the format for packing the count is wrong? For offsetformat is set to '<Q' but the count should formatted as '<I'?

Good catch, using `offsetformat` here is wrong for NDPI. Use `tagformat2[:2]`. Anyway, it's more complicated for other cases...

**Comment by cgohlke** [comment]
I looked into this issue. It should be possible to support overwriting tag values in NDPI files when offsets and integer values do not exceed 32-bit range, which is the case for most files < 4 GB and some common cases for files > 4 GB. Full support for patching NDPI files requires changes outside the `TiffTag.overwrite` function, e.g. one needs to keep track of the location of the high bytes in the file for all tags...

What is your use case? Which tags do you need to change? Does that involve `LONG` values or new offsets > 32-bit?

**Comment by erikogabrielsson** [comment]
The use case would be anonymizing string tags by overwriting with another value of same length.

**Comment by cgohlke** [comment]
Partly implemented in v2022.7.28.
